pkg/cached/cache.go

```diff
@@ -0,0 +1,257 @@
+	}
+	return cacheResults
+}
+
+// Rerun if:
+//   - The last run resulted in an error
+//   - Any of the dependency previously returned an error
+//   - Any of the dependency just returned an error
+//   - Any of the dependency's etag changed
+func (c *merger[K, T, V]) needsRunning(results map[K]Result[T]) bool {
+	if c.cacheResults == nil {
+		return true
+	}
+	if c.result.Err != nil {
+		return true
+	}
+	if len(results) != len(c.cacheResults) {
+		panic(fmt.Errorf("invalid number of results: %v (expected %v)", len(results), len(c.cacheResults)))
+	}
+	for key, oldResult := range c.cacheResults {
+		newResult, ok := results[key]
+		if !ok {
+			panic(fmt.Errorf("unknown cache entry: %v", key))
+		}
+
+		if newResult.Etag != oldResult.Etag || newResult.Err != nil || oldResult.Err != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *merger[K, T, V]) Get() Result[V] {
+	cacheResults := c.prepareResults()
+	if c.needsRunning(cacheResults) {
+		c.cacheResults = cacheResults
+		c.result = c.mergeFn(c.cacheResults)
+	}
+	return c.result
+}
+
+const transformerCacheName = "transformer"
+
+// NewTransformer creates a new cache that transforms the result of
+// another cache. The transformFn will only be called if the source
+// cache has updated the output, otherwise, the cached result will be
+// returned.
+//
+// If the dependency returned an error before, or it returns an error
+// this time, or if the transformerFn failed before, the function is
+// reran.
+func NewTransformer[T, V any](transformerFn func(Result[T]) Result[V], source Data[T]) Data[V] {
+	return NewMerger(func(caches map[string]Result[T]) Result[V] {
+		cache, ok := caches[transformerCacheName]
+		if len(caches) != 1 || !ok {
+			panic(fmt.Errorf("invalid cache for transformer cache: %v", caches))
+		}
+		return transformerFn(cache)
+	}, map[string]Data[T]{
+		transformerCacheName: source,
+	})
+}
+
+// NewSource creates a new cache that generates some data. This
+// will always be called since we don't know the origin of the data and
+// if it needs to be updated or not.
+func NewSource[T any](sourceFn func() Result[T]) Data[T] {
+	c := source[T](sourceFn)
+	return &c
+}
+
+type source[T any] func() Result[T]
+
+func (c *source[T]) Get() Result[T] {
+	return (*c)()
+}
+
+// NewStaticSource creates a new cache that always generates the
+// same data. This will only be called once.
+func NewStaticSource[T any](staticFn func() Result[T]) Data[T] {
+	return &static[T]{
+		fn: staticFn,
+	}
+}
+
+type static[T any] struct {
+	fn     func() Result[T]
+	result *Result[T]
+}
+
+func (c *static[T]) Get() Result[T] {
+	if c.result == nil {
+		result := c.fn()
+		c.result = &result
+	}
+	return *c.result
+}
+
+// Replaceable is a cache that carries the result even when the
+// cache is replaced. The cache can be replaced atomically (without any
+// lock held). This is the type that should typically be stored in
+// structs.
+type Replaceable[T any] struct {
+	cache  atomic.Value
```

Comment:
switch to `atomic.Pointer[Data[T]]`?

Reply:
Oh yeah, love it!

Reply:
Looks like this doesn't work with go1.18? reverted back to `atomic.Value`